File: typescript/all-your-base/all-your-base.test.ts
import { convert } from "./all-your-base";

describe("Converter", () => {
  it("single bit one to decimal", () => {
    expect(convert([1], 2, 10)).toEqual([1]);
  });

  it("binary to single decimal", () => {
    expect(convert([1, 0, 1], 2, 10)).toEqual([5]);
  });

  it("single decimal to binary", () => {
    expect(convert([5], 10, 2)).toEqual([1, 0, 1]);
  });

  it("binary to multiple decimal", () => {
    expect(convert([1, 0, 1, 0, 1, 0], 2, 10)).toEqual([4, 2]);
  });

  it("decimal to binary", () => {
    expect(convert([4, 2], 10, 2)).toEqual([1, 0, 1, 0, 1, 0]);
  });

  it("trinary to hexadecimal", () => {
    expect(convert([1, 1, 2, 0], 3, 16)).toEqual([2, 10]);
  });

  it("hexadecimal to trinary", () => {
    expect(convert([2, 10], 16, 3)).toEqual([1, 1, 2, 0]);
  });

  it("15-bit integer", () => {
    expect(convert([3, 46, 60], 97, 73)).toEqual([6, 10, 45]);
  });

  it("empty list", () => {
    expect(() => {
      convert([], 2, 10);
    }).toThrowError("Input has wrong format");
  });

  it("single zero", () => {
    expect(convert([0], 10, 2)).toEqual([0]);
  });

  it("multiple zeros", () => {
    expect(() => {
      convert([0, 0, 0], 10, 2);
    }).toThrowError("Input has wrong format");
  });

  it("leading zeros", () => {
    expect(() => {
      convert([0, 6, 0], 7, 10);
    }).toThrowError("Input has wrong format");
  });

  it("negative digit", () => {
    expect(() => {
      convert([1, -1, 1, 0, 1, 0], 2, 10);
    }).toThrowError("Input has wrong format");
  });

  it("invalid positive digit", () => {
    expect(() => {
      convert([1, 2, 1, 0, 1, 0], 2, 10);
    }).toThrowError("Input has wrong format");
  });

  it("first base is one", () => {
    expect(() => {
      convert([], 1, 10);
    }).toThrowError("Wrong input base");
  });

  it("second base is one", () => {
    expect(() => {
      convert([1, 0, 1, 0, 1, 0], 2, 1);
    }).toThrowError("Wrong output base");
  });

  it("first base is zero", () => {
    expect(() => {
      convert([], 0, 10);
    }).toThrowError("Wrong input base");
  });

  it("second base is zero", () => {
    expect(() => {
      convert([7], 10, 0);
    }).toThrowError("Wrong output base");
  });

  it("first base is negative", () => {
    expect(() => {
      convert([1], -2, 10);
    }).toThrowError("Wrong input base");
  });

  it("second base is negative", () => {
    expect(() => {
      convert([1], 2, -7);
    }).toThrowError("Wrong output base");
  });

  it("both bases are negative", () => {
    expect(() => {
      convert([1], -2, -7);
    }).toThrowError("Wrong input base");
  });

  it("wrong output_base base not integer", () => {
    expect(() => {
      convert([0], 3, 2.5);
    }).toThrowError("Wrong output base");
  });
});

File: scala/all-your-base/src/test/scala/AllYourBaseTest.scala
import org.scalatest.{Matchers, FunSuite}

/** @version 2.3.0 */
class AllYourBaseTest extends FunSuite with Matchers {

  test("single bit one to decimal") {
    AllYourBase.rebase(2, List(1), 10) should be(Some(List(1)))
  }

  test("binary to single decimal") {
    AllYourBase.rebase(2, List(1, 0, 1), 10) should be(Some(List(5)))
  }

  test("single decimal to binary") {
    AllYourBase.rebase(10, List(5), 2) should be(Some(List(1, 0, 1)))
  }

  test("binary to multiple decimal") {
    AllYourBase.rebase(2, List(1, 0, 1, 0, 1, 0), 10) should be(
      Some(List(4, 2)))
  }

  test("decimal to binary") {
    AllYourBase.rebase(10, List(4, 2), 2) should be(
      Some(List(1, 0, 1, 0, 1, 0)))
  }

  test("trinary to hexadecimal") {
    AllYourBase.rebase(3, List(1, 1, 2, 0), 16) should be(Some(List(2, 10)))
  }

  test("hexadecimal to trinary") {
    AllYourBase.rebase(16, List(2, 10), 3) should be(Some(List(1, 1, 2, 0)))
  }

  test("15-bit integer") {
    AllYourBase.rebase(97, List(3, 46, 60), 73) should be(Some(List(6, 10, 45)))
  }

  test("empty list") {
    AllYourBase.rebase(2, List(), 10) should be(Some(List(0)))
  }

  test("single zero") {
    AllYourBase.rebase(10, List(0), 2) should be(Some(List(0)))
  }

  test("multiple zeros") {
    AllYourBase.rebase(10, List(0, 0, 0), 2) should be(Some(List(0)))
  }

  test("leading zeros") {
    AllYourBase.rebase(7, List(0, 6, 0), 10) should be(Some(List(4, 2)))
  }

  test("input base is one") {
    AllYourBase.rebase(1, List(0), 10) should be(None)
  }

  test("input base is zero") {
    AllYourBase.rebase(0, List(), 10) should be(None)
  }

  test("input base is negative") {
    AllYourBase.rebase(-2, List(1), 10) should be(None)
  }

  test("negative digit") {
    AllYourBase.rebase(2, List(1, -1, 1, 0, 1, 0), 10) should be(None)
  }

  test("invalid positive digit") {
    AllYourBase.rebase(2, List(1, 2, 1, 0, 1, 0), 10) should be(None)
  }

  test("output base is one") {
    AllYourBase.rebase(2, List(1, 0, 1, 0, 1, 0), 1) should be(None)
  }

  test("output base is zero") {
    AllYourBase.rebase(10, List(7), 0) should be(None)
  }

  test("output base is negative") {
    AllYourBase.rebase(2, List(1), -7) should be(None)
  }

  test("both bases are negative") {
    AllYourBase.rebase(-2, List(1), -7) should be(None)
  }
}
